### PR TITLE
Enable ft.aggregate with KNN

### PR DESCRIPTION
**Pull request by alonre24**
Adjust `ft.aggregate` pipeline so it will work appropriately with KNN queries that behave similarly to SORTBY+LIMIT step. For that, we check whether a query contains KNN sub-string, and add an artificial SORTER to the pipeline if we confirm that KNN section exists after parsing the query.

Moreover, we make sure that all pipeline steps will run locally in the coordinator after the KNN step is executed in the shard (as was done for other SORTBY steps).

### CODE REVIEW

File: coord/src/dist_aggregate.c [hunk 1]
```diff
@@ -535,10 +536,26 @@ void RSExecDistAggregate(RedisModuleCtx *ctx, RedisModuleString **argv, int argc
 
   int profileArgs = parseProfile(argv, argc, r);
   if (profileArgs == -1) goto err;
-
```
Comment (GuyAv46) [inline coord/src/dist_aggregate.c:538]: ?

File: coord/src/dist_plan.cpp [hunk 2]
```diff
@@ -396,10 +396,11 @@ int AGGPLN_Distribute(AGGPlan *src, QueryError *status) {
         break;
       }
       case PLN_T_ARRANGE: {
+        PLN_ArrangeStep *astp = (PLN_ArrangeStep *)current;
         // If we already had an arrange step, we can't distribute the second one
         if (!hadArrange) {
           hadArrange = true;
-          PLN_ArrangeStep *astp = (PLN_ArrangeStep *)current;
```
Comment (GuyAv46) [inline coord/src/dist_plan.cpp:402]: why is this moved?

File: coord/src/module.c [hunk 3]
```diff
@@ -508,24 +457,56 @@ static int rscParseProfile(searchRequestCtx *req, RedisModuleString **argv) {
...
-void prepareOptionalTopKCase(searchRequestCtx *req, RedisModuleString **argv, int argc, QueryError *status) {
+
+void setKNNSpecialCase(searchRequestCtx *req, specialCaseCtx *knn_ctx) {
+  if(!req->specialCases) {
+    req->specialCases = array_new(specialCaseCtx*, 1);
+  }
+  req->specialCases = array_append(req->specialCases, knn_ctx);
+  // Default: No SORTBY is given, or SORTBY is given by other field
+  // When first sorting by different field, the topk vectors should be passed to the coordinator heap
+  knn_ctx->knn.shouldSort = true;
```
Comment (GuyAv46) [inline coord/src/module.c:468]: This is relevant for `FT.SEARCH` only right?

File: src/aggregate/aggregate_plan.h [hunk 4]
```diff
@@ -162,6 +162,7 @@ struct AGGPlan {
   PLN_ArrangeStep *arrangement;
   PLN_FirstStep firstStep_s;  // Storage for initial plan
   uint64_t steptypes;         // Mask of step-types contained in plan
+  bool hasKnn;                // An indicator that the plan contains a KNN arrange step.
```
Comment (GuyAv46) [inline src/aggregate/aggregate_plan.h:165]: I'm not sure I understand why we need this attribute in the plan. Why don't mark the arrange step as "local only" and make it more general? This might be useful for other fields in the future

File: src/aggregate/aggregate_plan.c [hunk 5]
```diff
@@ -138,6 +138,22 @@ PLN_ArrangeStep *AGPLN_GetArrangeStep(AGGPlan *pln) {
...
+  newStp->base.type = PLN_T_ARRANGE;
+  newStp->base.dtor = arrangeDtor;
+  dllist_prepend(&pln->steps, &newStp->base.llnodePln);
+
+  newStp->isLimited = 1;
+  newStp->limit = k;
+  newStp->sortKeys = array_new(const char *, 1);
+  newStp->sortKeys = array_append(newStp->sortKeys, distFieldName);
+  newStp->sortAscMap = SORTASCMAP_INIT;  // all ascending which is the default
+  pln->hasKnn = true;
```
Comment (GuyAv46) [inline src/aggregate/aggregate_plan.c:152]: Why is this a general flag?

File: src/aggregate/aggregate_plan.c [hunk 6]
```diff
@@ -138,6 +138,22 @@ PLN_ArrangeStep *AGPLN_GetArrangeStep(AGGPlan *pln) {
   return NULL;
 }
 
+PLN_ArrangeStep *AGPLN_AddKNNArrangeStep(AGGPlan *pln, size_t k, const char *distFieldName) {
```
Comment (GuyAv46) [inline src/aggregate/aggregate_plan.c:141]: consider making this function more general (mostly rename to something like `AGPLN_AddArrangeStepByField` and give it `limit` instead of `k`)

File: src/query.c [hunk 7]
```diff
@@ -908,10 +908,6 @@ static IndexIterator *Query_EvalGeometryNode(QueryEvalCtx *q, QueryNode *node) {
 
 
 static IndexIterator *Query_EvalVectorNode(QueryEvalCtx *q, QueryNode *qn) {
-  if((q->reqFlags & QEXEC_F_IS_EXTENDED)) {
-    QueryError_SetErrorFmt(q->status, QUERY_EAGGPLAN, "VSS is not yet supported on FT.AGGREGATE");
-    return NULL;
-  }
```
Comment (GuyAv46) [inline src/query.c:914]: :partying_face: :champagne:

File: tests/pytests/test_aggregate.py [hunk 8]
```diff
@@ -1032,3 +1032,50 @@ def testGroupAfterSort(env):
     #    one for (t == AAAA, n == 0), one for (t == BBBB, n == 0), and one for (t == ????, n == 1)
 
     env.assertEqual(res, expected)
+
+
+def testWithKNN(env):
```
Comment (GuyAv46) [inline tests/pytests/test_aggregate.py:1037]: consider adding a test with no `SORTBY`, and a hybrid case. Also maybe a `RANGE` sanity?

File: tests/pytests/test_vecsim.py [hunk 9]
```diff
@@ -1182,21 +1182,37 @@ def test_hybrid_query_cosine():
...
+    for i in range(2, 11, 2):
+        conn.execute_command("HSET", f'doc{i}{{3}}', "v", create_np_array_typed([i] * dim).tobytes(), 'n', f'{11-i}')
+
+    # Expect both queries to return doc1, doc2 and doc3, as these are the closest 3 documents in terms of
+    # the vector fields, and the ones with distance lower than 10.
+    expected_res = [['dist', '1'], ['dist', '4'], ['dist', '9']]
+
+    for query in ["*=>[KNN 3 @v $BLOB]=>{$yield_distance_as: dist}",
+                  "@v:[VECTOR_RANGE 10 $BLOB]=>{$yield_distance_as: dist}"]:
+        res = conn.execute_command("FT.AGGREGATE", "idx", query, 'SORTBY', '1', '@dist',
```
Comment (GuyAv46) [inline tests/pytests/test_vecsim.py:1205]: This `SORTBY` is an unnecessary overhead as the results should be already sorted, and we should not encourage doing so. Consider removing this step from the `AGGREGATE` tests

**Review by GuyAv46 — COMMENTED** [review]
Exciting!

File: coord/src/module.c (hunk 3 above)
Comment (alonre24) [reply coord/src/module.c:468]: yes

File: src/aggregate/aggregate_plan.h [hunk 10]
```diff
@@ -96,6 +96,7 @@ typedef struct {
   int isLimited;                  // Flag if `LIMIT` keyword was used.
   uint64_t offset;                // Seek results. If 0, then no paging is applied
   uint64_t limit;                 // Number of rows to output
+  bool runLocal;                  // Indicator that this step should run only local (not in shards)
```
Comment (GuyAv46) [inline src/aggregate/aggregate_plan.h:99]: Nitpicking: consider putting this after `isLimited`, and changing it to `bool` as well to save on the struct size

**Review by GuyAv46 — APPROVED** [review]

File: tests/pytests/test_aggregate.py [hunk 11]
```diff
@@ -1032,3 +1032,59 @@ def testGroupAfterSort(env):
...
+    conn.execute_command('HSET', 'doc4{3}', 'v', create_np_array_typed([1] * dim).tobytes(), 'n', '200')
+    conn.execute_command('HSET', 'doc5{3}', 'v', create_np_array_typed([2] * dim).tobytes(), 'n', '200')
+    conn.execute_command('HSET', 'doc6{3}', 'v', create_np_array_typed([3] * dim).tobytes(), 'n', '200')
+
+    expected_res = [['n', '100', 'c', '1'], ['n', '200', 'c', '1']]
+    res = conn.execute_command('FT.AGGREGATE', 'idx', '*=>[KNN 2 @v $blob]=>{$yield_distance_as: dist}',
+                               'GROUPBY', '1', '@n',
+                               'REDUCE', 'COUNT', '0', 'AS', 'c', 'SORTBY', '1', '@n',
+                               'PARAMS', '2', 'blob', create_np_array_typed([0] * dim).tobytes(), 'DIALECT', '2')
+    env.assertEqual(res[1:], expected_res)
```
Comment (DvirDukhan) [inline tests/pytests/test_aggregate.py:1090]: nice test
please add parameters sanity test (e.g. errors during parse etc..)

File: tests/pytests/test_vecsim.py [hunk 12]
```diff
@@ -1182,21 +1182,43 @@ def test_hybrid_query_cosine():
         conn.execute_command('FT.DROPINDEX', 'idx', 'DD')
 
 
-def test_fail_ft_aggregate():
+def test_ft_aggregate_basic():
```
Comment (DvirDukhan) [inline tests/pytests/test_vecsim.py:1185]: can we do for all algos?

File: tests/pytests/test_aggregate.py (hunk 11 above)
Comment (alonre24) [reply tests/pytests/test_aggregate.py:1111]: Add to test_search_params

File: src/result_processor.c [hunk 13]
```diff
@@ -412,13 +412,6 @@ static int rpsortNext_innerLoop(ResultProcessor *rp, SearchResult *r) {
     }
     // we need to allocate a new result for the next iteration
     self->pooledResult = rm_calloc(1, sizeof(*self->pooledResult));
-    // collected `limit` results. No need to continue.
-    if (self->quickExit && self->pq->count == self->pq->size) {
-      rp->Next = rpsortNext_Yield;
-      return rpsortNext_Yield(rp, r);
-    }
-    // we need to allocate a new result for the next iteration
-    self->pooledResult = rm_calloc(1, sizeof(*self->pooledResult));
```
Comment (DvirDukhan) [inline src/result_processor.c]: why the removal here?

File: tests/pytests/test_vecsim.py (hunk 9 above)
Comment (DvirDukhan) [reply tests/pytests/test_vecsim.py:1213]: what about this comment?

File: src/result_processor.c (hunk 13 above)
Comment (GuyAv46) [reply src/result_processor.c]: merge instead of rebase...

File: tests/pytests/test_vecsim.py (hunk 9 above)
Comment (GuyAv46) [reply tests/pytests/test_vecsim.py:1213]: was addressed, now we sort in the test only on `RANGE` query

**Review by DvirDukhan — APPROVED** [review]
